open_spiel/games/battleship.cc

```diff
@@ -0,0 +1,639 @@
+  // order. Position i corresponds to the damage that the opponent's ship in
+  // position i of bs_game->configuration.ships has suffered.
+  std::vector<int> ship_damage(conf.ships.size(), 0);
+  // Since in general we might have repeated shots, we cannot simply increase
+  // the ship damage every time a shot hits a ship. For that, we keep track of
+  // whether a cell was already hit in the past. We reuse the
+  // serialization/deserialization routines for shots to map from (r, c) to cell
+  // index r * board_width + c.
+  std::vector<bool> cell_hit(conf.board_width * conf.board_height, false);
+
+  std::string information_state;
+  for (const auto& move : moves_) {
+    if (absl::holds_alternative<ShipPlacement>(move.action)) {
+      // The player observed *their own* ship placements.
+      if (move.player == player) {
+        absl::StrAppend(&information_state, "/");
+        absl::StrAppend(&information_state,
+                        absl::get<ShipPlacement>(move.action).ToString());
+      }
+    } else {
+      const Shot& shot = absl::get<Shot>(move.action);
+
+      if (move.player != player) {
+        // If the shot came from the opponent, the player has seen it.
+        absl::StrAppend(&information_state, "/oppshot_", shot.ToString());
+      } else {
+        const int cell_index = SerializeShotAction_(shot);
+
+        char shot_outcome = 'W';  // For 'water'.
+        for (int ship_index = 0; ship_index < conf.ships.size(); ++ship_index) {
+          const Ship& ship = conf.ships.at(ship_index);
+
+          // SAFETY: the call to FindShipPlacement_ is safe, because if we are
+          // here it means that all ships have been placed.
+          const ShipPlacement ship_placement =
+              FindShipPlacement_(ship, opponent);
+
+          if (ship_placement.CoversCell(shot)) {
+            if (!cell_hit[cell_index]) {
+              // This is a new hit: we have to increas the ship damage and mark
+              // the cell as already hit.
+              ++ship_damage.at(ship_index);
+              cell_hit.at(cell_index) = true;
+            }
+            if (ship_damage.at(ship_index) == ship.length) {
+              shot_outcome = 'S';  // For 'sunk'.
+            } else {
+              shot_outcome = 'H';  // For 'hit' (but not sunk).
+            }
+          }
+        }
+
+        // Otherwise, the player knows they shot, but also knows whether the
+        // shot hit the water, hit a ship (but did not sink it), or sank a
+        // ship.
+        absl::StrAppend(&information_state, "/shot_", shot.ToString(), ":");
+
+        // XXX(gfarina): Apparently appending chars to a string is not a thing
+        // in Abseil?
+        information_state += shot_outcome;
+      }
+    }
+  }
+
+  return information_state;
+}
+
+std::string BattleshipState::ObservationString(Player player) const {
+  return InformationStateString(player);
```

Comment:
Maybe also briefly restate the format of the info states here, because normally this is not true.

Reply:
I'm going to address this comment partially in the response to the comment below about `InformationStateString`.

Reply:
For observations we normally just encode what the agent sees right now or what just happened, without the full history. E.g. in Tic Tac Toe it is the current state of the board (9 characters). In Kuhn/Leduc it is just the last thing that happened (new information since the last time). This method is partly to provide observations in the way typically understood by RL applications.

This is the place where the imperfect recall would come in. For example in battleship you could encode which ships are damaged / sunk but forget the order the shots occurred.

Do you plan to use observation string if there is already information state string that does the same? If not then I think it's ok to just not provide it (also reflecting that in the game type).

Reply:
Perfect, I think I got it. I'll implement the correct version that does not distinguish between the order of the moves.

Do you think it would be desirable to have observation return something pretty (for example a grid of characters with newlines and everything, and probably a lot of whitespace really), or is it better to have a more raw string like what I did for InformationState?

Reply:
Does not matter as long as it satifies the purpose. Maybe the pretty one makes it easier to debug but less compact in terms of memory. I doubt the memory savings make much difference for the use cases so makes sense to make it nice looking but really does not matter that much, so feel free to do whatever is easier.